src/network/networkd-sysctl.c

```diff
@@ -40,10 +42,82 @@ static void manager_set_ip_forwarding(Manager *manager, int family) {
                                   enable_disable(t), af_to_ipv4_ipv6(family));
 }
 
+static void sysctl_overwrite_check(sd_event_source *source, const struct inotify_event *event, Hashmap *sysctls) {
+        _cleanup_free_ char *buf = NULL, *path = NULL;
+        const char *dir_path, *value;
+        size_t len;
+        int r;
+
+        assert(event);
+
+        r = sd_event_source_get_inotify_path(source, &dir_path);
+        if (r < 0)
+                return;
+
+        path = path_join(dir_path, event->name);
+        if (!path)
+                return;
+
+        value = hashmap_get(sysctls, path);
+
+        /* We don't handle this sysctl */
+        if (!value)
+                return;
+
+        if (read_full_file(path, &buf, &len) < 0)
+                return;
+
+        delete_trailing_chars(buf, "\n");
+
+        if (!streq(value, buf))
+                log_warning("sysctl %s was changed from '%s' to '%s'", path, value, buf);
+}
+
+static int sysctl_inotify_manager(sd_event_source *source, const struct inotify_event *event, void *userdata) {
+        Manager *manager = ASSERT_PTR(userdata);
+
+        sysctl_overwrite_check(source, event, manager->sysctls);
+
+        return 0;
+}
+
+static int sysctl_inotify_link(sd_event_source *source, const struct inotify_event *event, void *userdata) {
+        Link *link = ASSERT_PTR(userdata);
+
+        sysctl_overwrite_check(source, event, link->sysctls);
+
+        return 0;
+}
```

Comment:
sysctl_inotify_manager() and sysctl_inotify_link() are pretty much the same. maybe make them one, and then pass a pointer to the pointer of the hashmap as userdata instead of the link or manager object.

Reply:
When the two calls to sd_event_add_inotify() are done, the shadow hashmap is still NULL.
Passing the pointer directly won't work in this case, while passing manager or link do

Reply:
A solution could be to initialize the hashmap on manager and link creation, and use hashmap_replace() instead of hashmap_ensure_replace()

Reply:
that's why i suggested to pass a pointer to a pointer, i.e. pass the location of where the Hashmap* pointer will be placed, then you can initialize it lazily as currently.